Add common python linters and align codebase (+ minor touch-ups)

eyalfrish opened the pull request:
## What's new?
1. This PR adds a bunch of "basic" project linters commonly used with python projects. These were added using
the `pre-commit` dev-package (via `dev-requirements.txt`) and the linters configuration file (`.pre-commit-config.yaml`).
2. GitHub Action (`.github/workflows/pre-commit.yml`) support for pre-commit workflow (per push).
2. `.gitignore` beefed-up a bit, supporting common ignores (vens, python binaries, IDEs...).
3. Re-ordering (ABC-wise) of `requirements.txt`.
4. Addition of pre-commit usage in the `README.md`.
5. Adaptations of code to comply with linters (99% automatic).

## Included Linters
1. [Auto] **pre-commit hooks** - built-in hooks (<https://pre-commit.com/hooks.html>)
   * trailing-whitespace
   * end-of-file-fixer
   * check-added-large-files
2. [Partially Auto] **markdownlint** - *.md checker (<https://github.com/markdownlint/markdownlint>)
6. [Auto] **autoflake** - remove unused imports (<https://github.com/PyCQA/autoflake/>)
7. [Auto] **isort** - import sorter (<https://pycqa.github.io/isort/>)
8. [Auto] **black** - code style formatter (<https://black.readthedocs.io/en/stable/>)
9. **mypy** - type checker (<https://mypy-lang.org/>)

## Setting-up and Using the Pre-Commit Linters (Optional) -
#### Installing dev package:
`python -m pip install -r dev-requirements.txt`

#### Installing pre-commit hooks (after this, linters will run before commits):
`pre-commit install`

#### Running pre-commit hooks manually on all files:
`pre-commit run --all-files`

eyalfrish committed:
Update README.md with pre-commit usage

eyalfrish committed:
ABC-reordering of requirements.txt

eyalfrish committed:
Add pre-commit with basic linters

Linters include -
  1) trailing-whitespace (auto-fix)
  2) end-of-file-fixer (auto-fix)
  3) large-files warning
  4) markdownlint, *.md checker (partial auto-fix)
  4) autoflake, unused imports/vars removal (auto-fix)
  5) isort, imports sorting (auto-fix)
  6) black, strict styling (auto-fix)
  7) mypy, types checking

eyalfrish committed:
Beef-up .gitignroe with commonly useful ignores

eyalfrish committed:
Minor code-facelift (using typehints and slightly more pythonic)

eyalfrish committed:
Move supported-docs entirely into DocumentLoader

eyalfrish committed:
Add GitHub pre-commit action

eyalfrish committed:
Advancing versions of all linters (+ fix python version of isort)

eyalfrish committed:
Advancing versions of all GitHub Actions

sharonmordechai commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,45 @@
+  - repo: https://github.com/psf/black
+    rev: 24.4.0
+    hooks:
+      - id: black
```
What about adding the"pylint" hook as well?

eyalfrish replied:
Ok. Added pylint.

eyalfrish committed:
Add pylint to pre-commit and align code to comply

Main warnings addressed -
   1. Missing doc-strings
   2. Long lines
   3. Usage of "with" (handled differently)

While at it -
   1. Adding missing type-hints
   2. Fixed wrong temp_dir code (also a pylint issue)
   3. Suppressed a few pylint stuff

eyalfrish committed:
Bump-down linters' python version (3.11->3.9)

This is a temporary change

eyalfrish commented:
@sharonmordechai, as expected, i dragged into this MR a bit more than needed. Added type-hints (verified by `mypy`) a bit all over the code. Also, a minor face-lift to some parts.

Anyway, im still convincing (not too well) myself that its "ok" that this is all in one MR which is "basic tools/styling" to the repo.

Anyway, im done (for now). Please confirm its not broken, of course ;). Feel free to comment again, as you see fit! And, lets kill this MR :)!!!

eyalfrish committed:
Bump-down the workflow's python version (3.11.6->3.9.12)

This is a temporary change
